File: algorithms/02_c/dp/lis.c
#include <stdio.h>

#define MAXN 100000

int lis_length(const int *a, int n) {
    int tails[MAXN], len = 0;
    for (int i = 0; i < n; ++i) {
        int lo = 0, hi = len;
        while (lo < hi) {
            int m = (lo + hi) / 2;
            if (tails[m] < a[i]) lo = m + 1;
            else hi = m;
        }
        tails[lo] = a[i];
        if (lo == len) ++len;
    }
    return len;
}

int main(void) {
    int a[] = {10, 9, 2, 5, 3, 7, 101, 18};
    int n = sizeof(a) / sizeof(a[0]);

    if (lis_length(a, n) != 4) {
        printf("[C LIS] FAILED: length %d, want 4\n", lis_length(a, n));
        return 1;
    }
    printf("[C LIS] Longest increasing subsequence verified (O(n log n))\n");
    return 0;
}

File: algorithms/06_javascript/dp/lis.js
function lengthOfLIS(nums) {
  const tails = [];
  for (const x of nums) {
    let lo = 0, hi = tails.length;
    while (lo < hi) {
      const mid = (lo + hi) >> 1;
      if (tails[mid] < x) lo = mid + 1;
      else hi = mid;
    }
    if (lo === tails.length) tails.push(x);
    else tails[lo] = x;
  }
  return tails.length;
}

module.exports = { lengthOfLIS };

if (require.main === module) {
  if (lengthOfLIS([10, 9, 2, 5, 3, 7, 101, 18]) !== 4) throw new Error("LIS failed");
  if (lengthOfLIS([0, 1, 0, 3, 2, 3]) !== 4) throw new Error("LIS failed");
  console.log("[JavaScript LIS] O(n log n) longest increasing subsequence verified");
}
